textutils-2.1/man/sum.1
.\" DO NOT MODIFY THIS FILE!  It was generated by help2man 1.28.
.TH SUM "1" "July 2002" "sum (textutils) 2.1" "User Commands"
.SH NAME
sum \- checksum and count the blocks in a file
.SH SYNOPSIS
.B sum
[\fIOPTION\fR]... [\fIFILE\fR]...
.SH DESCRIPTION
.\" Add any additional description here
.PP
Print checksum and block counts for each FILE.
.TP
\fB\-r\fR
defeat \fB\-s\fR, use BSD sum algorithm, use 1K blocks
.TP
\fB\-s\fR, \fB\-\-sysv\fR
use System V sum algorithm, use 512 bytes blocks
.TP
\fB\-\-help\fR
display this help and exit
.TP
\fB\-\-version\fR
output version information and exit
.PP
With no FILE, or when FILE is -, read standard input.
.SH AUTHOR
Written by Kayvan Aghaiepour and David MacKenzie.
.SH "REPORTING BUGS"
Report bugs to <bug-textutils@gnu.org>.
.SH COPYRIGHT
Copyright \(co 2002 Free Software Foundation, Inc.
.br
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
.SH "SEE ALSO"
The full documentation for
.B sum
is maintained as a Texinfo manual.  If the
.B info
and
.B sum
programs are properly installed at your site, the command
.IP
.B info sum
.PP
should give you access to the complete manual.


zsh/ChangeLog
beta 0.75
Wed Jun 14 14:47:46 2000    <amold@SKYNYRD>
	* ntport\dirent.c: Fix serious bug in wildcard expansion on Win9x 
beta 0.74
Mon Jan 31 14:53:06 2000    <amold@SKYNYRD>
	* ntport\support.c: fixed bug in converting '\'s in $HOME 
	to '/'s.
beta 0.73
Mon Jan 17 11:55:15 2000    <amold@SKYNYRD>
	* ntport\support.c: fixed crash in long arglist introduced by
 	0.70
beta 0.72
Tue Jan 11 08:28:29 2000    <amold@SKYNYRD>
	* ntport\support.c: Fixed memory corruption on Win9x
beta 0.71
Mon Jan 10 11:49:39 2000    <amold@SKYNYRD>
	* ntport\stdio.c: Fix bug in inheritance of file handles.  
beta 0.69,0.70
Fri Dec 17 11:53:19 1999    <amold@SKYNYRD>
	* ntport\support.c: Fixed bug in argv[0] of spawned argument. Might
	have manifested as applications in ZSHROOT failing to spawn 
	themselves. 
	Also added winntnoquoteprotect option. Use with tools like MKS
	if you see unexpected quoting of "
beta 0.68
Mon Sep 27 11:20:24 1999    <amold@SKYNYRD>
	* ntport\support.c: support paths > 1024 characters. 
beta 0.67
Tue Sep 07 10:05:02 1999    <amold@SKYNYRD>
	* ntport\io.c: Fix bug that affected FPATH/autoload 
beta 0.65
Fri Aug 27 11:41:58 1999    <amold@SKYNYRD>
	* ntport\dirent.c: Fix completion with drive letters in string. 
beta 0.64
Fri Aug 27 11:04:11 1999    <amold@SKYNYRD>
	* ntport\io.c: Unfix Win95 spurious linefeed fix for Windows 98. 
beta 0.63
Wed May 19 16:21:37 1999    <amold@SKYNYRD>
	* exec.c, misc: VC 6.0 compliation errors. Disable optmizations
	in functions that call fork() to avoid losing global variables
	that might be in registers.
beta 0.62
Tue Apr 13 17:06:12 1999    <amold@SKYNYRD>
	* ntport\support.c, zle_misc.c: Keep server name in prompt
 	if %C is specified and pwd is UNC.
beta 0.61
Tue Mar 23 12:17:50 1999  unknown  <bogus@SKYNYRD>
	* ntport\support.c: bug fix for previous fix. 
beta 0.60
Sat Feb 13 17:49:14 1999    <amold@SKYNYRD>
	* ntport\support.c,ntbin.c: Fix lame arg-concatenation in
 	exec and start paths. commands with large argument strings should
	now execute *much* faster.
beta 0.57
Tue Feb 02 18:09:39 1999    <amold@SKYNYRD>
	* ntport\ntbin.c: fixed bug that improperly passed arguments
	to applications launched by windows "shell" associations. 
	For example, .bat files.
beta 0.55
Mon Oct 26 11:09:44 1998    <amold@SKYNYRD>
	* ntport\ntbin.c: start builtin will not use '/' for option
 	flags. only '-' will work.
beta 0.54
Fri Oct 16 12:52:08 1998    <amold@SKYNYRD>
	* ntport\support.c: added WINNTWAITFORGUIAPPS option. 
beta 0.53
Mon Oct 12 12:31:59 1998    <amold@SKYNYRD>
	* init.c: look for "zshrc" as well as ".zshrc" 
beta 0.52
Mon Sep 28 17:31:08 1998    <amold@SKYNYRD>
	* utils.c: attempt to ignore case while correcting spelling
 	if WINNTIGNORECASE is set.
beta 0.51
Thu Sep 17 13:48:27 1998    <amold@SKYNYRD>
	* hashtable.c,support.c: filenames without extensions and
 	*only* files with extensions specified in the PATHEXT variable
 	will be hashed. The default extensions are ".com;.exe;.cmd;.bat"
beta 0.50
Wed Sep 16 15:58:56 1998    <amold@SKYNYRD>
	* hashtable.c: added hashing for external exes without
	extension. bogus spelling corrections should be reduced.
beta 0.49
Tue Sep 15 11:09:31 1998    <amold@SKYNYRD>
	* support.c: The shell will automatically set TERM to vt100
 	if it's not set in the environment. There should be no need
	to set the variable explicitly any more.
beta 0.48
Mon Aug 03 17:47:08 1998    <amold@SKYNYRD>
	* support.c: Fixed bug in converting paths to '/'-delimited 
beta 0.47
Mon Aug 03 16:59:38 1998    <amold@SKYNYRD>
	* utils.c: Fixed bug in applying previous patch. If %~ wasn't
	working, this should fix it.
beta 0.46
Sat Jul 18 14:36:10 1998    <amold@SKYNYRD>
	* utils.c,globals.h,zsh.h: patch from dominik for 
	winntconvertbackslash.
beta 0.45
Mon Jun 29 11:30:55 1998    <amold@SKYNYRD>
	* utils.c: patch from dominik vogt for honoring winntignorecase
 	in %~ prompt.
beta 0.44
Mon Jun 22 11:19:12 1998    <amold@SKYNYRD>
	* support.c,io.c: moved some startup code around. fixed spurious
 	linefeeds on win95
beta 0.43
Tue Jun 09 17:25:29 1998    <amold@SKYNYRD>
	* builtin.c,ntport\io.c: Fix for the read builtin not
 	echoing characters
beta 0.429
Thu Jun 04 17:43:54 1998    <amold@SKYNYRD>
	* utils.c: Fixed %~ prompt spec.
beta 0.428
Thu Jun 04 16:50:21 1998    <amold@SKYNYRD>
	* builtin.c: Fixed problems with the setting of PWD.
beta 0.427
Thu May 28 14:31:28 1998    <amold@SKYNYRD>
	* ntport\support.c: Fix to let apps with spaces in their path
	run. c:\program files\myapp.exe , for example.
beta 0.426
Wed May 27 10:46:10 1998    <amold@SKYNYRD>
	* ntport\support.c: Fixed stupid command buffer overflow
 	problem. If you crashed while grepping through a bunch of
	files, this should fix it.
beta 0.425
Fri May 22 08:55:10 1998    <amold@SKYNYRD>
	* ntport\support.c: Fixed bug in parsing args to child processes.
beta 0.424
Sun May 17 13:18:21 1998    <amold@SKYNYRD>
	* ntport\support.c: Fixed bug that caused zsh to spawn a new
	console if $path contained a component '/'. 
beta 0.423
Fri May 15 16:08:17 1998    <amold@SKYNYRD>
	* ntport\ntbin.c: winntlamepathfix now works with the 'start'
	builtin.
beta 0.421
Mon Apr 06 16:18:20 1998    <amold@SKYNYRD>
	* ntport\support.c,fork.c: Fixed bug in heap initialization for
	forked children that crashed the child process.
beta 0.42
Sun Apr 05 13:36:49 1998    <amold@SKYNYRD>
	* ntport\support.c: Put in code to fix the case of a directory
 	you cd into. This is extremely expensive and really silly, so
	it may become optional in the future.
beta 0.41
Sat Apr 04 15:27:10 1998    <amold@SKYNYRD>
	* ntport\support.c: Fixed bug that restricted command line
 	length
beta 0.401
Thu Mar 26 08:21:21 1998    <amold@SKYNYRD>
	* ntport\support.c: Fix to stop home from being re-set in child
 	if unset in parent. Also HOME will not be set by zsh if ZDOTDIR

textutils-2.1/src/sum.c
/* sum -- checksum and count the blocks in a file
   Copyright (C) 86, 89, 91, 1995-2001 Free Software Foundation, Inc.

   This program is free software; you can redistribute it and/or modify
   it under the terms of the GNU General Public License as published by
   the Free Software Foundation; either version 2, or (at your option)
   any later version.

   This program is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.

   You should have received a copy of the GNU General Public License
   along with this program; if not, write to the Free Software Foundation,
   Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.  */

/* Like BSD sum or SysV sum -r, except like SysV sum if -s option is given. */

/* Written by Kayvan Aghaiepour and David MacKenzie. */

#include <config.h>

#include <stdio.h>
#include <sys/types.h>
#include <getopt.h>
#include "system.h"
#include "closeout.h"
#include "error.h"
#include "human.h"
#include "safe-read.h"

/* The official name of this program (e.g., no `g' prefix).  */
#define PROGRAM_NAME "sum"

#define AUTHORS N_ ("Kayvan Aghaiepour and David MacKenzie")

/* The name this program was run with. */
char *program_name;

/* Nonzero if any of the files read were the standard input. */
static int have_read_stdin;

static struct option const longopts[] =
{
  {"sysv", no_argument, NULL, 's'},
  {GETOPT_HELP_OPTION_DECL},
  {GETOPT_VERSION_OPTION_DECL},
  {NULL, 0, NULL, 0}
};

void
usage (int status)
{
  if (status != 0)
    fprintf (stderr, _("Try `%s --help' for more information.\n"),
	     program_name);
  else
    {
      printf (_("\
Usage: %s [OPTION]... [FILE]...\n\
"),
	      program_name);
      fputs (_("\
Print checksum and block counts for each FILE.\n\
\n\
  -r              defeat -s, use BSD sum algorithm, use 1K blocks\n\
  -s, --sysv      use System V sum algorithm, use 512 bytes blocks\n\
"), stdout);
      fputs (HELP_OPTION_DESCRIPTION, stdout);
      fputs (VERSION_OPTION_DESCRIPTION, stdout);
      fputs (_("\
\n\
With no FILE, or when FILE is -, read standard input.\n\
"), stdout);
      printf (_("\nReport bugs to <%s>.\n"), PACKAGE_BUGREPORT);
    }
  exit (status == 0 ? EXIT_SUCCESS : EXIT_FAILURE);
}

/* Calculate and print the rotated checksum and the size in 1K blocks
   of file FILE, or of the standard input if FILE is "-".
   If PRINT_NAME is >1, print FILE next to the checksum and size.
   The checksum varies depending on sizeof(int).
   Return 0 if successful, -1 if an error occurs. */

static int
bsd_sum_file (const char *file, int print_name)
{
  register FILE *fp;
  register int checksum = 0;	/* The checksum mod 2^16. */
  register uintmax_t total_bytes = 0;	/* The number of bytes. */
  register int ch;		/* Each character read. */
  char hbuf[LONGEST_HUMAN_READABLE + 1];

  if (STREQ (file, "-"))
    {
      fp = stdin;
      have_read_stdin = 1;
    }
  else
    {
      fp = fopen (file, "rb");
      if (fp == NULL)
	{
	  error (0, errno, "%s", file);
	  return -1;
	}
    }
  /* Need binary I/O, or else byte counts and checksums are incorrect.  */
  SET_BINARY (fileno(fp));

  while ((ch = getc (fp)) != EOF)
    {
      total_bytes++;
      checksum = (checksum >> 1) + ((checksum & 1) << 15);
      checksum += ch;
      checksum &= 0xffff;	/* Keep it within bounds. */
    }

  if (ferror (fp))
    {
      error (0, errno, "%s", file);
      if (!STREQ (file, "-"))
	fclose (fp);
      return -1;
    }

  if (!STREQ (file, "-") && fclose (fp) == EOF)
    {
      error (0, errno, "%s", file);
      return -1;
    }

  printf ("%05d %5s", checksum,
	  human_readable_inexact (total_bytes, hbuf, 1, 1024, human_ceiling));
  if (print_name > 1)
    printf (" %s", file);
  putchar ('\n');

  return 0;
}

/* Calculate and print the checksum and the size in 512-byte blocks
   of file FILE, or of the standard input if FILE is "-".
   If PRINT_NAME is >0, print FILE next to the checksum and size.
   Return 0 if successful, -1 if an error occurs. */

static int
sysv_sum_file (const char *file, int print_name)
{
  int fd;
  unsigned char buf[8192];
  register int bytes_read;
  uintmax_t total_bytes = 0;
  char hbuf[LONGEST_HUMAN_READABLE + 1];
  int r;
  int checksum;

  /* The sum of all the input bytes, modulo (UINT_MAX + 1).  */
  register unsigned int s = 0;

  if (STREQ (file, "-"))
    {
      fd = 0;
      have_read_stdin = 1;
    }
  else
    {
      fd = open (file, O_RDONLY| O_BINARY);
      if (fd == -1)
	{
	  error (0, errno, "%s", file);
	  return -1;
	}
    }
  /* Need binary I/O, or else byte counts and checksums are incorrect.  */
  SET_BINARY (fd);

  while ((bytes_read = safe_read (fd, buf, sizeof buf)) > 0)
    {
      register int i;

      for (i = 0; i < bytes_read; i++)
	s += buf[i];
      total_bytes += bytes_read;
    }

  if (bytes_read < 0)
    {
      error (0, errno, "%s", file);
      if (!STREQ (file, "-"))
	close (fd);
      return -1;
    }

  if (!STREQ (file, "-") && close (fd) == -1)
    {
      error (0, errno, "%s", file);
      return -1;
    }

  r = (s & 0xffff) + ((s & 0xffffffff) >> 16);
  checksum = (r & 0xffff) + (r >> 16);

  printf ("%d %s", checksum,
	  human_readable_inexact (total_bytes, hbuf, 1, 512, human_ceiling));
  if (print_name)
    printf (" %s", file);
  putchar ('\n');

  return 0;
}

int
main (int argc, char **argv)
{
  int errors = 0;
  int optc;
  int files_given;
  int (*sum_func) PARAMS ((const char *, int)) = bsd_sum_file;

  program_name = argv[0];
  setlocale (LC_ALL, "");
  bindtextdomain (PACKAGE, LOCALEDIR);
  textdomain (PACKAGE);

  atexit (close_stdout);

  have_read_stdin = 0;

  while ((optc = getopt_long (argc, argv, "rs", longopts, NULL)) != -1)
    {
      switch (optc)
	{
	case 0:
	  break;

	case 'r':		/* For SysV compatibility. */
	  sum_func = bsd_sum_file;
	  break;

	case 's':
	  sum_func = sysv_sum_file;
	  break;

	case_GETOPT_HELP_CHAR;

	case_GETOPT_VERSION_CHAR (PROGRAM_NAME, AUTHORS);

	default:
	  usage (1);
	}
    }

  files_given = argc - optind;
  if (files_given == 0)
    {
      if ((*sum_func) ("-", files_given) < 0)
	errors = 1;
    }
  else
    for (; optind < argc; optind++)
      if ((*sum_func) (argv[optind], files_given) < 0)
	errors = 1;

  if (have_read_stdin && fclose (stdin) == EOF)
    error (EXIT_FAILURE, errno, "-");
  exit (errors == 0 ? EXIT_SUCCESS : EXIT_FAILURE);
}
